L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+    auto lessSkippedLayers = [](State* lhs, State* rhs) { return lhs->numSkippedLayers() < rhs->numSkippedLayers(); };
+    stable_sort(stream.begin(), stream.end(), lessSkippedLayers);
+    // sort in number of consistent stubs
+    auto moreConsistentLayers = [](State* lhs, State* rhs) { return lhs->numConsistentLayers() > rhs->numConsistentLayers(); };
+    stable_sort(stream.begin(), stream.end(), moreConsistentLayers);
+    // sort in track id
+    stable_sort(stream.begin(), stream.end(), [](State* lhs, State* rhs){ return lhs->trackId() < rhs->trackId(); });
+    // keep first state (best due to previous sorts) per track id
+    stream.erase(unique(stream.begin(), stream.end(), [](State* lhs, State* rhs){ return lhs->track() == rhs->track(); }), stream.end());
+    //for (State* state : stream)
+      //cout << state->x0() << " " << state->x1() << " " << dataFormats_->chosenRofPhi() << endl;
+  }
+
+  // updates state
+  void KalmanFilter::update(State*& state) {
+    // R-Z plane
+    const double H12 = H12_->digi(state->H12());
+    const double m1 = m1_->digi(state->m1());
+    const double v1 = v1_->digi(state->v1());
+    double x2 = x2_->digi(state->x2());
+    double x3 = x3_->digi(state->x3());
+    double C22 = C22_->digi(state->C22());
+    double C23 = C23_->digi(state->C23());
+    double C33 = C33_->digi(state->C33());
+    v1_->updateRangeActual(v1);
+    C22_->updateRangeActual(C22);
+    C23_->updateRangeActual(C23);
+    C33_->updateRangeActual(C33);
+    H12_->updateRangeActual(H12);
+    m1_->updateRangeActual(m1);
+    const double r1C = x3_->digi(m1  - x3);
+    const double r1 = r1_->digi(r1C - x2 * H12);
+    const double S12 = S12_->digi(C23 + H12 * C22);
+    const double S13 = S13_->digi(C33 + H12 * C23);
+    const double R11C = S13_->digi(v1 + S13);
+    const double R11 = R11_->digi(R11C + H12 * S12);
+    // imrpoved dynamic cancelling
+    const int msbZ = max(0, (int)ceil(log2(R11 / R11_->base())));
+    const double R11Rough = R11Rough_->digi(R11 * pow(2., 16 - msbZ));
+    const double invR11Approx = invR11Approx_->digi(1. / R11Rough);
+    const double invR11Cor = invR11Cor_->digi(2. - invR11Approx * R11Rough);
+    const double invR11 = invR11_->digi(invR11Approx * invR11Cor * pow(2., 16 - msbZ));
+    const double K21 = K21_->digi(S12 * invR11);
+    const double K31 = K31_->digi(S13 * invR11);
+    x2 = x2_->digi(x2 + r1 * K21);
+    x3 = x3_->digi(x3 + r1 * K31);
+    C22 = C22_->digi(C22 - S12 * K21);
+    C23 = C23_->digi(C23 - S13 * K21);
+    C33 = C33_->digi(C33 - S13 * K31);
+    // R-Phi plane
+    const double H00 = H00_->digi(state->H00());
+    const double m0 = m0_->digi(state->m0());
+    const double v0 = v0_->digi(state->v0());
+    double x0 = x0_->digi(state->x0());
+    double x1 = x1_->digi(state->x1());
+    double C00 = C00_->digi(state->C00());
+    double C01 = C01_->digi(state->C01());
+    double C11 = C11_->digi(state->C11());
+    v0_->updateRangeActual(v0);
```

Comment:
Do the calls to updateRangeActual() have no effect on the KF, and are only used in the endJob to recommend changes in the ranges? If so, add comment saying this.